File: solutions/beecrowd/1157/1157.clj
(ns main)

(defn divisible? [number divisor]
  (zero? (rem number divisor)))

(defn all-divisors [number]
  (loop [possible-divisor 1
         found-divisors []]
    (if (<= possible-divisor (int (Math/sqrt number)))
      (if (divisible? number possible-divisor)
        (recur (inc possible-divisor) (concat found-divisors [possible-divisor (/ number possible-divisor)]))
        (recur (inc possible-divisor) found-divisors))
      found-divisors)))

(defn main []
  (loop [line (read-line)]
    (when line
      (doseq [number (sort (set (all-divisors (Integer/parseInt line))))]
        (println number))
      (recur (read-line)))))

(main)

File: solutions/beecrowd/1157/1157.py
import heapq
import math
import sys

for line in sys.stdin:
    number = int(line)
    divisors = []
    found_divisors = set()

    for possible_divisor in range(1, int(math.sqrt(number)) + 1):
        if number % possible_divisor == 0:
            found_divisors.add(possible_divisor)
            heapq.heappush(divisors, possible_divisor)

            if number // possible_divisor not in found_divisors:
                found_divisors.add(number // possible_divisor)
                heapq.heappush(divisors, number // possible_divisor)

    while divisors:
        print(heapq.heappop(divisors))
